File: library/src/main/java/com/snail/commons/utils/SystemUtils.java
package com.snail.commons.utils;

import android.app.ActivityManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.pm.ActivityInfo;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.ServiceInfo;
import android.location.LocationManager;
import android.os.Build;
import android.os.Environment;
import android.os.PowerManager;
import android.os.StatFs;
import android.os.storage.StorageManager;
import android.provider.Settings;

import com.snail.commons.entity.Storage;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.reflect.Array;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;


/**
 * 安卓系统的工具类
 *
 * @author Zeng
 */
public class SystemUtils {
    /**
     * 判断位置服务是否打开
     */
    public static boolean isLocationEnabled(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            LocationManager locationManager = (LocationManager) context.getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
            if (locationManager != null) {
                return locationManager.isLocationEnabled();
            }
        } else {
            try {
                int locationMode = Settings.Secure.getInt(context.getContentResolver(), Settings.Secure.LOCATION_MODE);
                return locationMode != Settings.Secure.LOCATION_MODE_OFF;
            } catch (Settings.SettingNotFoundException e) {
                e.printStackTrace();
            }
        }        
        return false;
    }

    /**
     * 判断GPS是否打开
     */
    public static boolean isGPSEnabled(Context context) {
        LocationManager locationManager = (LocationManager) context.getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
        return locationManager != null && locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
    }

    /**
     * 判断服务是否正在运行
     *
     * @param context   上下文
     * @param className 服务的完整类名
     */
    public static boolean isServiceRunning(Context context, String className) {
        ActivityManager am = (ActivityManager) context.getApplicationContext().getSystemService(Context.ACTIVITY_SERVICE);
        if (am == null) {
            return false;
        }
        List<ActivityManager.RunningServiceInfo> services = am.getRunningServices(200);
        for (ActivityManager.RunningServiceInfo serviceInfo : services) {
            if (serviceInfo.service.getClassName().equals(className)) {
                return true;
            }
        }
        return false;
    }

    /**
     * 判断屏幕是否亮着
     * @param context 上下文
     */
    public static boolean isScreenOn(Context context) {
        PowerManager pm = (PowerManager) context.getApplicationContext().getSystemService(Context.POWER_SERVICE);
        if (pm == null) {
            return false;
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT_WATCH) {
            return pm.isInteractive();
        } else {
            return pm.isScreenOn();
        }
    }

    /**
     * 判断当前系统是否安装指定的应用
     * @param context 上下文
     * @param packageName 要判断的应用包名
     */
    public boolean isAppInstalled(Context context, String packageName) {
        PackageManager packageManager = context.getPackageManager();
        // 获取所有已安装程序的包信息
        List<PackageInfo> pinfo = packageManager.getInstalledPackages(0);
        for (int i = 0; i < pinfo.size(); i++) {
            if(pinfo.get(i).packageName.equals(packageName)) {
                return true;
            }
        }
        return false;
    }
    
    /**
     * 获取总内存大小，单位是byte
     */
    public static long getTotalMemSize() {
        try {
            BufferedReader br = new BufferedReader(new InputStreamReader(new FileInputStream("/proc/meminfo")));
            char[] info = br.readLine().toCharArray();
            StringBuilder sb = new StringBuilder();
            for (char c : info) {
                if (c >= '0' && c <= '9') {
                    sb.append(c);
                }
            }
            long kbSize = Long.parseLong(sb.toString());
            br.close();
            return (kbSize * 1024);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return 0;
    }

    /**
     * 获取可用内存大小，单位byte
     */
    public static long getAvailMemSize(Context context) {
        ActivityManager am = (ActivityManager) context.getApplicationContext().getSystemService(Context.ACTIVITY_SERVICE);
        ActivityManager.MemoryInfo outInfo = new ActivityManager.MemoryInfo();
        if (am == null) {
            return -1;
        }
        am.getMemoryInfo(outInfo);
        return outInfo.availMem;
    }

    /**
     * 获取正在运行的进程数
     */
    public static int getRunningProcessCount(Context context) {
        ActivityManager am = (ActivityManager) context.getApplicationContext().getSystemService(Context.ACTIVITY_SERVICE);
        if (am == null) {
            return -1;
        }
        List<ActivityManager.RunningAppProcessInfo> runningAppProcesses = am.getRunningAppProcesses();
        return runningAppProcesses == null ? 0 : runningAppProcesses.size();
    }

    /**
     * 扩展卡是否可用
     */
    public static boolean isSdCardAvailable() {
        return Environment.MEDIA_MOUNTED.equals(Environment.getExternalStorageState());
    }

    /**
     * 获取手机内部剩余存储空间
     *
     * @return 字节数
     */
    public static long getInternalFreeSpace() {
        File path = Environment.getDataDirectory();
        StatFs stat = new StatFs(path.getPath());
        return stat.getAvailableBlocksLong() * stat.getBlockSizeLong();
    }

    /**
     * 获取扩展卡剩余存储空间
     *
     * @return 字节数。
     */
    public static long getExternalFreeSpace() {
        if (isSdCardAvailable()) {
            File path = Environment.getExternalStorageDirectory();
            StatFs stat = new StatFs(path.getPath());
            return stat.getAvailableBlocksLong() * stat.getBlockSizeLong();
        } else {
            return 0;
        }
    }

    /**
     * 获取存储卡剩余大小
     */
    public static long getStorageFreeSpace(String path) {
        if (path != null) {
            File file = new File(path);
            if (file.exists()) {
                StatFs stat = new StatFs(path);
                return stat.getAvailableBlocksLong() * stat.getBlockSizeLong();
            }
        }
        return 0;
    }

    /**
     * 存储卡总容量
     */
    public static long getStorageTotalSpace(String path) {
        if (path != null) {
            File file = new File(path);
            if (file.exists()) {
                StatFs stat = new StatFs(path);
                return stat.getBlockSizeLong() * stat.getBlockCountLong();
            }
        }
        return 0;
    }

    /**
     * 获取所有存储路径
     */
    public static List<String> getStoragePaths(Context context) {
        StorageManager sm = (StorageManager) context.getApplicationContext().getSystemService(Context.STORAGE_SERVICE);
        try {
            String[] volumePaths = (String[]) sm.getClass().getMethod("getVolumePaths").invoke(sm);
            List<String> pathList = new ArrayList<>();
            for (String path : volumePaths) {
                if (getStorageTotalSpace(path) > 0) {
                    pathList.add(path);
                }
            }
            return pathList;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public static ArrayList<Storage> getStorages(Context context) {
        StorageManager storageManager = (StorageManager) context.getApplicationContext().getSystemService(Context.STORAGE_SERVICE);
        try {
            //得到StorageManager中的getVolumeList()方法的对象
            Method getVolumeList = storageManager.getClass().getMethod("getVolumeList");
            //得到StorageVolume类的对象
            Class<?> storageValumeClazz = Class.forName("android.os.storage.StorageVolume");
            //获得StorageVolume中的一些方法
            Method getPath = storageValumeClazz.getMethod("getPath");
            Method isRemovable = storageValumeClazz.getMethod("isRemovable");
            Method allowMassStorage = storageValumeClazz.getMethod("allowMassStorage");
            Method primary = storageValumeClazz.getMethod("isPrimary");
            Method description = storageValumeClazz.getMethod("getDescription", Context.class);

            Method mGetState = null;
            //getState 方法是在4.4_r1之后的版本加的，之前版本（含4.4_r1）没有
            // （http://grepcode.com/file/repository.grepcode.com/java/ext/com.google.android/android/4.4_r1/android/os/Environment.java/）
            if (Build.VERSION.SDK_INT > Build.VERSION_CODES.KITKAT) {
                try {
                    mGetState = storageValumeClazz.getMethod("getState");
                } catch (NoSuchMethodException e) {
                    e.printStackTrace();
                }
            }

            //调用getVolumeList方法，参数为：“谁”中调用这个方法
            Object invokeVolumeList = getVolumeList.invoke(storageManager);
            int length = Array.getLength(invokeVolumeList);
            ArrayList<Storage> list = new ArrayList<>();
            for (int i = 0; i < length; i++) {
                Object storageValume = Array.get(invokeVolumeList, i);//得到StorageVolume对象
                String path = (String) getPath.invoke(storageValume);
                boolean removable = (boolean) isRemovable.invoke(storageValume);
                boolean isAllowMassStorage = (boolean) allowMassStorage.invoke(storageValume);
                boolean isPrimary = (boolean) primary.invoke(storageValume);
                String desc = (String) description.invoke(storageValume, context);
                String state;
                if (mGetState != null) {
                    state = (String) mGetState.invoke(storageValume);
                } else {
                    state = Environment.getStorageState(new File(path));
                }
                long totalSize = 0;
                long availaleSize = 0;
                if (Environment.MEDIA_MOUNTED.equals(state)) {
                    totalSize = getStorageTotalSpace(path);
                    availaleSize = getStorageFreeSpace(path);
                }
                Storage storage = new Storage();
                storage.availaleSize = availaleSize;
                storage.totalSize = totalSize;
                storage.state = state;
                storage.path = path;
                storage.isRemovable = removable;
                storage.description = desc;
                storage.isAllowMassStorage = isAllowMassStorage;
                storage.isPrimary = isPrimary;
                storage.isUsb = desc != null && desc.toLowerCase().contains("usb");
                list.add(storage);
            }
            return list;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * 存储器是否被挂载
     */
    public static boolean isMounted(Context context, String path) {
        StorageManager sm = (StorageManager) context.getApplicationContext().getSystemService(Context.STORAGE_SERVICE);
        try {
            String state = (String) sm.getClass().getMethod("getVolumeState", String.class).invoke(sm, path);
            return Environment.MEDIA_MOUNTED.equals(state);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    public static String getSystemProperty(String propName) {
        String line;
        BufferedReader input = null;
        try {
            Process p = Runtime.getRuntime().exec("getprop " + propName);
            input = new BufferedReader(new InputStreamReader(p.getInputStream()), 1024);
            line = input.readLine();
            input.close();
        } catch (IOException ex) {
            return null;
        } finally {
            if (input != null) {
                try {
                    input.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return line;
    }

    /**
     * 获取Application的Meta值
     *
     * @param context 上下文
     * @param name    meta名
     * @return 没有返回null
     */
    public static String getApplicationMetaValue(Context context, String name) {
        try {
            ApplicationInfo info = context.getPackageManager().getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);
            Object value = info.metaData.get(name);
            return value == null ? null : value.toString();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * 获取Activity的Meta值
     *
     * @param context 上下文
     * @param cls     Activity的class
     * @param name    meta名
     * @return 没有返回null
     */
    public static String getActivityMetaValue(Context context, Class<?> cls, String name) {
        try {
            ActivityInfo info = context.getPackageManager().getActivityInfo(new ComponentName(context, cls.getName()), PackageManager.GET_META_DATA);
            Object value = info.metaData.get(name);
            return value == null ? null : value.toString();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * 获取Receiver的Meta值
     *
     * @param context 上下文
     * @param cls     Receiver的class
     * @param name    meta名
     * @return 没有返回null
     */
    public static String getReceiverMetaValue(Context context, Class<?> cls, String name) {
        try {
            ActivityInfo info = context.getPackageManager().getReceiverInfo(new ComponentName(context, cls.getName()), PackageManager.GET_META_DATA);
            Object value = info.metaData.get(name);
            return value == null ? null : value.toString();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * 获取Service的Meta值
     *
     * @param context 上下文
     * @param cls     Service的class
     * @param name    meta名
     * @return 没有返回null
     */
    public static String getServiceMetaValue(Context context, Class<?> cls, String name) {
        try {
            ServiceInfo info = context.getPackageManager().getServiceInfo(new ComponentName(context, cls.getName()), PackageManager.GET_META_DATA);
            Object value = info.metaData.get(name);
            return value == null ? null : value.toString();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: library/src/main/java/com/snail/commons/utils/SystemUtils.kt
package com.snail.commons.utils

import android.app.ActivityManager
import android.content.ComponentName
import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.location.LocationManager
import android.os.Build
import android.os.Environment
import android.os.PowerManager
import android.os.StatFs
import android.os.storage.StorageManager
import android.provider.Settings
import androidx.core.os.EnvironmentCompat
import com.snail.commons.entity.Storage
import java.io.*
import java.lang.reflect.Method
import java.util.*


/**
 * 安卓系统的工具类
 *
 * @author Zeng
 */
object SystemUtils {

    /**
     * 获取总内存大小，单位是byte
     */
    @JvmStatic
    val totalMemSize: Long
        get() {
            try {
                val br = BufferedReader(InputStreamReader(FileInputStream("/proc/meminfo")))
                val info = br.readLine().toCharArray()
                val sb = StringBuilder()
                for (c in info) {
                    if (c in '0'..'9') {
                        sb.append(c)
                    }
                }
                val kbSize = java.lang.Long.parseLong(sb.toString())
                br.close()
                return kbSize * 1024
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return 0
        }

    /**
     * 扩展卡是否可用
     */
    @JvmStatic
    val isSdCardAvailable: Boolean
        get() = Environment.MEDIA_MOUNTED == Environment.getExternalStorageState()

    /**
     * 获取手机内部剩余存储空间
     *
     * @return 字节数
     */
    @JvmStatic
    val internalFreeSpace: Long
        get() {
            val path = Environment.getDataDirectory()
            val stat = StatFs(path.path)
            return stat.availableBlocksLong * stat.blockSizeLong
        }

    /**
     * 获取扩展卡剩余存储空间
     *
     * @return 字节数。
     */
    @JvmStatic
    val externalFreeSpace: Long
        get() {
            return if (isSdCardAvailable) {
                val stat = StatFs(Environment.getExternalStorageDirectory().absolutePath)
                stat.availableBlocksLong * stat.blockSizeLong
            } else {
                0
            }
        }

    /**
     * 获取存储卡剩余大小
     */
    @JvmStatic
    fun getStorageFreeSpace(path: String): Long {
        if (File(path).exists()) {
            val stat = StatFs(path)
            return stat.availableBlocksLong * stat.blockSizeLong
        }
        return 0
    }

    /**
     * 存储卡总容量
     */
    @JvmStatic
    fun getStorageTotalSpace(path: String): Long {
        if (File(path).exists()) {
            val stat = StatFs(path)
            return stat.blockSizeLong * stat.blockCountLong
        }
        return 0
    }    

    @JvmStatic
    fun getSystemProperty(propName: String): String? {
        val line: String
        var input: BufferedReader? = null
        try {
            val p = Runtime.getRuntime().exec("getprop $propName")
            input = BufferedReader(InputStreamReader(p.inputStream), 1024)
            line = input.readLine()
            input.close()
        } catch (ex: IOException) {
            return null
        } finally {
            if (input != null) {
                try {
                    input.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
        return line
    }
}

/*############################################ 扩展函数 #########################################*/

/**
 * 判断当前系统是否安装指定的应用
 * @param packageName 要判断的应用包名
 */
fun Context.isAppInstalled(packageName: String): Boolean {
    // 获取所有已安装程序的包信息
    val pinfo = packageManager.getInstalledPackages(0)
    for (i in pinfo.indices) {
        if (pinfo[i].packageName == packageName) {
            return true
        }
    }
    return false
}

/**
 * 判断位置服务是否打开
 */
fun Context.isLocationEnabled(): Boolean {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
        val locationManager = applicationContext.getSystemService(Context.LOCATION_SERVICE) as? LocationManager
        if (locationManager != null) {
            return locationManager.isLocationEnabled
        }
    } else {
        try {
            val locationMode = Settings.Secure.getInt(contentResolver, Settings.Secure.LOCATION_MODE)
            return locationMode != Settings.Secure.LOCATION_MODE_OFF
        } catch (e: Settings.SettingNotFoundException) {
            e.printStackTrace()
        }
    }
    return false
}

/**
 * 判断GPS是否打开
 */
fun Context.isGPSEnabled(): Boolean {
    val locationManager = applicationContext.getSystemService(Context.LOCATION_SERVICE) as? LocationManager
    return locationManager != null && locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
}

/**
 * 判断屏幕是否亮着
 */
fun Context.isScreenOn(): Boolean {
    return (applicationContext.getSystemService(Context.POWER_SERVICE) as? PowerManager)?.isInteractive == true
}

/**
 * 获取可用内存大小，单位byte
 */
fun Context.getAvailMemSize(): Long {
    val am = applicationContext.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager
    val outInfo = ActivityManager.MemoryInfo()
    if (am == null) {
        return -1
    }
    am.getMemoryInfo(outInfo)
    return outInfo.availMem
}

/**
 * 获取正在运行的进程数
 */
fun Context.getRunningProcessCount(): Int {
    val am = applicationContext.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager ?: return -1
    return am.runningAppProcesses?.size ?: 0
}

/**
 * 获取所有存储路径
 */
fun Context.getStoragePaths(): List<String>? {
    return try {
        val sm = applicationContext.getSystemService(Context.STORAGE_SERVICE) as StorageManager
        val volumePaths = sm.javaClass.getMethod("getVolumePaths").invoke(sm) as Array<String>
        val pathList = ArrayList<String>()
        for (path in volumePaths) {
            if (SystemUtils.getStorageTotalSpace(path) > 0) {
                pathList.add(path)
            }
        }
        pathList
    } catch (e: Exception) {
        null
    }
}

/**
 * 获取设备存储信息
 */
fun Context.getStorages(): ArrayList<Storage>? {
    return try {
        val storageManager = applicationContext.getSystemService(Context.STORAGE_SERVICE) as StorageManager
        //得到StorageManager中的getVolumeList()方法的对象
        val getVolumeList = storageManager.javaClass.getMethod("getVolumeList")
        //得到StorageVolume类的对象
        val storageValumeClazz = Class.forName("android.os.storage.StorageVolume")
        //获得StorageVolume中的一些方法
        val getPath = storageValumeClazz.getMethod("getPath")
        val isRemovable = storageValumeClazz.getMethod("isRemovable")
        val allowMassStorage = storageValumeClazz.getMethod("allowMassStorage")
        val primary = storageValumeClazz.getMethod("isPrimary")
        val description = storageValumeClazz.getMethod("getDescription", Context::class.java)

        var mGetState: Method? = null
        try {
            mGetState = storageValumeClazz.getMethod("getState")
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        }

        //调用getVolumeList方法，参数为：“谁”中调用这个方法
        val invokeVolumeList = getVolumeList.invoke(storageManager)
        val length = java.lang.reflect.Array.getLength(invokeVolumeList)
        val list = ArrayList<Storage>()
        for (i in 0 until length) {
            val storageValume = java.lang.reflect.Array.get(invokeVolumeList, i)//得到StorageVolume对象
            val path = getPath.invoke(storageValume) as? String ?: ""
            val removable = isRemovable.invoke(storageValume) as? Boolean ?: false
            val isAllowMassStorage = allowMassStorage.invoke(storageValume) as? Boolean ?: false
            val isPrimary = primary.invoke(storageValume) as? Boolean ?: false
            val desc = description.invoke(storageValume, this) as? String ?: ""
            val state = if (mGetState != null) {
                mGetState.invoke(storageValume) as? String
            } else {
                Environment.getStorageState(File(path))
            }
            var totalSize: Long = 0
            var availaleSize: Long = 0
            if (Environment.MEDIA_MOUNTED == state) {
                totalSize = SystemUtils.getStorageTotalSpace(path)
                availaleSize = SystemUtils.getStorageFreeSpace(path)
            }
            val storage = Storage()
            storage.availaleSize = availaleSize
            storage.totalSize = totalSize
            storage.state = state ?: EnvironmentCompat.MEDIA_UNKNOWN
            storage.path = path
            storage.isRemovable = removable
            storage.description = desc
            storage.isAllowMassStorage = isAllowMassStorage
            storage.isPrimary = isPrimary
            storage.isUsb = desc.toLowerCase(Locale.ENGLISH).contains("usb")
            list.add(storage)
        }
        list
    } catch (e: Exception) {
        null
    }
}

/**
 * 存储器是否被挂载
 */
fun Context.isMounted(path: String): Boolean {
    return try {
        val sm = applicationContext.getSystemService(Context.STORAGE_SERVICE) as StorageManager
        val state = sm.javaClass.getMethod("getVolumeState", String::class.java).invoke(sm, path) as String
        Environment.MEDIA_MOUNTED == state
    } catch (e: Exception) {
        false
    }
}

/**
 * 判断apk是否是debug包
 */
fun Context.isDebugApk(apkPath: String): Boolean {
    val info = packageManager.getPackageArchiveInfo(apkPath, PackageManager.GET_ACTIVITIES)
    return try {
        (info.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0
    } catch (e: Exception) {
        false
    }
}

/**
 * 判断app是否运行在debug模式下
 */
fun Context.isRunInDebug(): Boolean {
    return try {
        (applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0
    } catch (e: Exception) {
        false
    }
}

/**
 * 获取Service的Meta值
 *
 * @param cls     Service的class
 * @param name    meta名
 * @return 没有返回null
 */
fun Context.getServiceMetaValue(cls: Class<*>, name: String): String? {
    return try {
        val info = packageManager.getServiceInfo(ComponentName(this, cls.name), PackageManager.GET_META_DATA)
        val value = info.metaData.get(name)
        value?.toString()
    } catch (e: Exception) {
        null
    }
}

/**
 * 获取Receiver的Meta值
 *
 * @param cls     Receiver的class
 * @param name    meta名
 * @return 没有返回null
 */
fun Context.getReceiverMetaValue(cls: Class<*>, name: String): String? {
    return try {
        val info = packageManager.getReceiverInfo(ComponentName(this, cls.name), PackageManager.GET_META_DATA)
        val value = info.metaData.get(name)
        value?.toString()
    } catch (e: Exception) {
        null
    }    
}

/**
 * 获取Activity的Meta值
 *
 * @param cls     Activity的class
 * @param name    meta名
 * @return 没有返回null
 */
fun Context.getActivityMetaValue(cls: Class<*>, name: String): String? {
    return try {
        val info = packageManager.getActivityInfo(ComponentName(this, cls.name), PackageManager.GET_META_DATA)
        val value = info.metaData.get(name)
        value?.toString()
    } catch (e: Exception) {
        null
    }
}

/**
 * 获取Application的Meta值
 *
 * @param name    meta名
 * @return 没有返回null
 */
fun Context.getApplicationMetaValue(name: String): String? {
    return try {
        val info = packageManager.getApplicationInfo(packageName, PackageManager.GET_META_DATA)
        val value = info.metaData.get(name)
        value?.toString()
    } catch (e: Exception) {
        null
    }
}